src/flag_attn/flash.py

```diff
@@ -45,14 +46,38 @@ def forward(ctx, q, k, v, causal, sm_scale):
                 num_warps=num_warps, num_stages=num_stages,
             )
 
+            if return_total_attention:
+                tot_attn = torch.empty((B, H, N), device=q.device, dtype=torch.float32)
+                grid = (triton.cdiv(N, BLOCK_N), H, B)
+                _total_attention_kernel[grid](
+                    q, k, L, tot_attn, sm_scale, 
+                    q.stride(0), q.stride(1), q.stride(2), q.stride(3),
+                    k.stride(0), k.stride(1), k.stride(2), k.stride(3),
+                    B, H, M, P_SEQ,
+                    BLOCK_M=BLOCK_M, BLOCK_DMODEL=D, BLOCK_N=BLOCK_N, 
+                    CAUSAL=causal,
+                    DIVISIBLE_M=divisible_m, DIVISIBLE_N=divisible_n,
+                    num_stages=num_stages, num_warps=num_warps,
+                )
+
         # autograd context maintenance
         ctx.save_for_backward(q, k, v, o, L)
         ctx.sm_scale = sm_scale
         ctx.causal = causal
-        return o
+
+        has_extra_return = return_log_normalizer or return_total_attention
+        if has_extra_return:
+            outs = (
+                o,
+                L if return_log_normalizer else None,
+                tot_attn if return_total_attention else None
+            )
```

Comment:
what about only one extra return is needed?

Reply:
I want to preserve the structure, which means to return the extra slots when any of the extra_outputs is requested.

When no extra outpus are requested, only return the attention outputs.
When any extra outputs is requested, return a 3-tuple. (attention_outputs, log_normalizer, total_attention). Though the extra-outputs that are not requested is `None`.